### PR TITLE
Remove Tasks and Payments

### LINKED ISSUE #1126: Check (and fix) interaction between One Tx Payment extension and a non-zero global claim delay
OneTxPayment uses expenditures for gas efficiency if there is more than one recipient, but we think it doesn't play nicely if there is a global claim delay (but it could/should).

### PR DESCRIPTION
**Pull request by kronosapiens**
Closes #1126 

Removes Tasks and Payments from the core contracts. Specifically:

- Deletes `ColonyTask.sol` and `ColonyPayments.sol` and removes all references to them from the rest of the contracts
- View functions remain, as does the ability to remove funds from existing Tasks and Payments
- Migrates the test suite away from Tasks and towards Expenditures as a way of giving out tokens and reputations
- For the reputation tests, migrates to direct reputation rewards and penalties

For simplicity, the tests have been left otherwise untouched, meaning that we still give out reputation in the same way as we did when we used Tasks (i.e. three recipients, penalties, etc).

A bit unrelated, but this PR also does the following:

- Updates the OneTxPayment extension to set the `globalClaimDelay` to 0

### CODE REVIEW

File: test/extensions/voting-rep.js [hunk 1]
```diff
@@ -1307,15 +1307,15 @@ contract("Voting Reputation", (accounts) => {
...
 
       await forwardTime(FAIL_EXECUTION_TIMEOUT_PERIOD, this);
 
       const { logs } = await voting.finalizeMotion(motionId);
       expect(logs[0].args.executed).to.be.false;
 
-      const tasksAfter = await colony.getTaskCount();
-      expect(tasksAfter).to.eq.BN(tasksBefore);
+      const balanceAfter = await token.balanceOf(colony.address);
+      expect(balanceAfter).to.eq.BN(balanceBefore);
```
[the remark this answers is not in the document]
Comment (kronosapiens) [reply test/extensions/voting-rep.js:1318]: It's a large codebase 😅

[the remark this answers is not in the document]
Comment (kronosapiens) [reply test/extensions/voting-rep.js:1318]: Seems like it was my fault: https://github.com/JoinColony/colonyNetwork/commit/d8d0917eca016c843f8270fdf008324a5f1e33ae

**Comment by area** [comment]
Out of a meeting I've just had with Arren/Jack: OneTxPayment should work even if there is a non-zero global claim delay for expenditures.

**Comment by kronosapiens** [comment]
Sounds simple enough, we just add arbitration permission and zero out the claim delay on finalization.

File: helpers/test-data-generator.js [hunk 2]
```diff
@@ -36,221 +16,114 @@ const Resolver = artifacts.require("Resolver");
...
     evaluatorPayout,
     workerPayout,
-    managerRating,
-    workerRating,
   });
 
-  await colony.finalizeTask(taskId);
-  return taskId;
+  await colony.finalizeExpenditure(expenditureId, { from: manager });
+  await colony.claimExpenditurePayout(expenditureId, 0, tokenAddress);
```
[the remark this answers is not in the document]
Comment (kronosapiens) [reply helpers/test-data-generator.js:124]: I hear that. I suppose we could do a big find-replace, but I also see an argument for leaving as-is or fixing as we go. A big migration could introduce subtle errors which tie us up for a while. I'm sympathetic to a new developer experience, but I think there's also an argument that a new developer would learn the repo's history and know what they were looking at.

[the remark this answers is not in the document]
Comment (area) [reply helpers/test-data-generator.js:124]: Take a look at the `maint/the-cullings-slottening` branch and `cherry-pick` the commit in here if you think it's okay.

[the remark this answers is not in the document]
Comment (kronosapiens) [reply helpers/test-data-generator.js:124]: We do things like `USER0`, etc but those constants stand in for user addresses, not just integers

File: contracts/colony/Colony.sol [hunk 3]
```diff
@@ -400,4 +401,57 @@ contract Colony is BasicMetaTransaction, Multicall, ColonyStorage, PatriciaTreeP
     return tokenApprovalTotals[_token];
   }
 
+  // Deprecated view functions for Tasks and Payments
```
[the remark this answers is not in the document]
Comment (kronosapiens) [reply contracts/colony/Colony.sol:404]: It's only a few functions at this point, so I don't think it's worth the overhead of incorporating a new contract into the setup process. If we end up doing more deprecations then let's revisit this.

File: .gitignore [hunk 4]
```diff
@@ -26,3 +26,4 @@ truffle-security-output.json
 .coverage_contracts
 etherrouter-address.json
 ganache-chain-db
+colonyNetwork-glwss4
```
[the remark this answers is not in the document]
Comment (area) [reply .gitignore]: Actually, this shouldn't be in this PR at all - this is coming from #1164

**Comment by kronosapiens** [comment]
@area I've chosen to skip the `*Deprecated.sol` files for now, it's more overhead for contract compilation and deployment and I think the gains are minimal. If we continue deprecating large chunks of functionality we should revisit the idea.

**Review by area — COMMENTED** [review]
We should remove permissions around the functions being removed in `finishUpgrade`, as well as removing them from `ColonyAuthority` for future colonies.

I would like to keep the comments around deprecated structs / events, even if they're not going in to their own file.

**Comment by kronosapiens** [comment]
@area I've given this another review today, I think it's ready for your green light

File: contracts/colony/ColonyFunding.sol [hunk 5]
```diff
@@ -348,20 +267,12 @@ contract ColonyFunding is ColonyStorage { // ignore-swc-123
       );
     }
 
-    if (
-      fromPot.associatedType == FundingPotAssociatedType.Expenditure ||
-      fromPot.associatedType == FundingPotAssociatedType.Payment ||
-      fromPot.associatedType == FundingPotAssociatedType.Task
-    ) {
+    if (fromPot.associatedType == FundingPotAssociatedType.Expenditure) {
```
[the remark this answers is not in the document]
Comment (area) [reply contracts/colony/ColonyFunding.sol:315]: How should I interpret 'thumbs up, no change' :sweat_smile:

[the remark this answers is not in the document]
Comment (kronosapiens) [reply contracts/colony/ColonyFunding.sol:315]: Slipped by me

**Review by area — COMMENTED** [review]
What do you think of, in `finishUpgrade` of `OneTxPayment`:

* Check if we (the extension) has root permission in the colony.
* If not, no further actions and return
* If it does, award itself arbitration permission in root domain

This obviously doesn't capture all scenarios, but will mean that the front-end team don't have to worry about handling the adding of the permission on upgrade, based on how the extension is currently used.

**Comment by kronosapiens** [comment]
@area works for me. Do you want me to add it or you?

**Comment by area** [comment]
> What do you think of, in `finishUpgrade` of `OneTxPayment`:
> 
>     * Check if we (the extension) has root permission in the colony.
> 
>     * If not, no further actions and return
> 
>     * If it does, award itself arbitration permission in root domain
> 
> 
> This obviously doesn't capture all scenarios, but will mean that the front-end team don't have to worry about handling the adding of the permission on upgrade, based on how the extension is currently used.

This doesn't actually work, OneTxPayment is only awarded Funding and Administration currently, so I'll remove the commit.

**Review by area — APPROVED** [review]

**Pull request merged**